### PR TITLE
evm.md: Remove #callWithCode, dont copy the entire program to K cell

### PR DESCRIPTION
**Pull request by MrChico**
As programs can get quite large, we don't want to be copying them in their full glory to the K cell. We can still get the bytecode from the accounts cell.

### CODE REVIEW

**Comment by MrChico** [comment]
I've already signed the cla. Did something change?

**Comment by dwightguth** [comment]
This isn't actually going to save any memory because K uses immutable representations of terms, so a "copy" like this is really just a pointer write.

**Comment by ehildenb** [comment]
I agree with @dwightguth that this probably isn't necessary.

Also, if your email changed or something, or your GitHub account changed, then the CLA will not pick up on it. And CLA you sign for K or some other repo won't apply here.

I would have to see some profiling numbers to demonstrate that this actually reduces memory usage to be convinced. Post them here.

You can use GNU `time` (not the shell builtin `time`) to get rough profiling numbers.

**Comment by MrChico** [comment]
Hmm, I didn't know about this pointer situation... It came up as it ends up as a quite a large printed term with redundant parts. Especially since we're already omitting the program and programBytes cell. Of course we could probably just omit the `#callWithCode` and `#mkCall` production, but it still doesn't seem as clean as this solution to me

File: evm.md [hunk 1]
```diff
@@ -1219,15 +1219,13 @@ The various `CALL*` (and other inter-contract control flow) operations will be d
...
 -   `#return__` is a placeholder for the calling program, specifying where to place the returned data in memory.
 
 ```k
     syntax InternalOp ::= "#checkCall" Int Int
                         | "#call"         Int Int Int Int Int WordStack Bool
-                        | "#callWithCode" Int Int Map WordStack Int Int WordStack Bool
-                        | "#mkCall"       Int Int Map WordStack     Int WordStack Bool
- // ----------------------------------------------------------------------------------
+                        | "#mkCall"       Int Int Int Int WordStack Bool
+ // ------------------------------------------------------------------------
```
Comment (ehildenb) [inline evm.md]: formatting

Comment (ehildenb) [reply evm.md]: line up the arguments of `#call` and `#mkCall`, make the horizontal rule the correct length.

File: evm.md [hunk 2]
```diff
@@ -1297,10 +1309,13 @@ The various `CALL*` (and other inter-contract control flow) operations will be d
...
          <caller> _ => ACCTFROM </caller>
-         <program> _ => CODE </program>
-         <programBytes> _ => BYTES </programBytes>
+         <program> _ => .Map </program>
+         <programBytes> _ => .WordStack </programBytes>
          <static> OLDSTATIC:Bool => OLDSTATIC orBool STATIC </static>
          <touchedAccounts> ... .Set => SetItem(ACCTFROM) SetItem(ACCTTO) ... </touchedAccounts>
+         <activeAccounts> ACCTS </activeAccounts>
+         <schedule> SCHED </schedule>
+      requires notBool ACCTCODE in #precompiledAccounts(SCHED) andBool notBool ACCTCODE in ACCTS
```
Comment (ehildenb) [inline evm.md]: Put the `andBool` on the next line, and line up the `l` in `andBool` with the `s` in `requires`.

**Comment by ehildenb** [comment]
Performance numbers are in (run twice to verify).
Entries are `exit-code  wall-time  user-time  sys-time  peak-mem-usage`:

```
* 06c521b5 - evm.md: Remove #callWithCode, dont copy the entire program to K cell             - Martin Lundfall (HEAD -> tested-profile, mrchico/omitProgram, omitProgram)
|  0  0.10s    0.06s     0.04s    16052KB     bash  -c  make  clean
|  0  131.07s  217.43s   12.04s   819512KB    bash  -c  make  llvm-deps         -B
|  0  55.54s   418.82s   15.76s   2148676KB   bash  -c  make  build-java        build-llvm                   build-ocaml  -B  -j6
|  0  0.12s    0.09s     0.03s    15996KB     bash  -c  make  split-tests       -B
|  0  233.58s  2119.77s  881.36s  4488676KB   bash  -c  make  test-conformance  TEST_CONCRETE_BACKEND=ocaml  -j16
|  0  142.68s  1686.49s  460.99s  15364804KB  bash  -c  make  test-conformance  TEST_CONCRETE_BACKEND=llvm   -j16
|  0  577.29s  9646.41s  539.18s  2965680KB   bash  -c  make  test-proof        -j10
|
|  0  0.10s    0.05s     0.04s    15988KB     bash  -c  make  clean
|  0  129.72s  217.66s   11.78s   668196KB    bash  -c  make  llvm-deps         -B
|  0  55.88s   424.27s   16.91s   2350680KB   bash  -c  make  build-java        build-llvm                   build-ocaml  -B  -j6
|  0  0.13s    0.10s     0.03s    15880KB     bash  -c  make  split-tests       -B
|  0  234.12s  2121.52s  885.24s  4488516KB   bash  -c  make  test-conformance  TEST_CONCRETE_BACKEND=ocaml  -j16
|  0  142.60s  1686.00s  462.20s  15364836KB  bash  -c  make  test-conformance  TEST_CONCRETE_BACKEND=llvm   -j16
|  0  567.45s  9755.07s  527.60s  3030440KB   bash  -c  make  test-proof        -j10
|
* 39c92c4e - deps/k: update submodule (#383)                                                  - rv-jenkins
|  0  0.07s    0.07s     0.00s    16000KB     bash  -c  make  clean
|  0  128.00s  215.02s   12.25s   829168KB    bash  -c  make  llvm-deps         -B
|  0  55.07s   418.79s   17.26s   2181924KB   bash  -c  make  build-java        build-llvm                   build-ocaml  -B  -j6
|  0  0.14s    0.11s     0.02s    15844KB     bash  -c  make  split-tests       -B
|  0  276.63s  2266.55s  886.37s  4488560KB   bash  -c  make  test-conformance  TEST_CONCRETE_BACKEND=ocaml  -j16
|  0  140.88s  1665.79s  454.69s  15365076KB  bash  -c  make  test-conformance  TEST_CONCRETE_BACKEND=llvm   -j16
|  0  564.88s  9499.99s  531.84s  3043472KB   bash  -c  make  test-proof        -j10
|
|  0  0.10s    0.06s     0.03s    15864KB     bash  -c  make  clean
|  0  128.07s  214.34s   11.99s   833492KB    bash  -c  make  llvm-deps         -B
|  0  54.65s   405.76s   15.76s   2375996KB   bash  -c  make  build-java        build-llvm                   build-ocaml  -B  -j6
|  0  0.14s    0.11s     0.03s    15956KB     bash  -c  make  split-tests       -B
|  0  276.68s  2267.15s  890.43s  4488732KB   bash  -c  make  test-conformance  TEST_CONCRETE_BACKEND=ocaml  -j16
|  0  141.49s  1671.64s  459.03s  15365476KB  bash  -c  make  test-conformance  TEST_CONCRETE_BACKEND=llvm   -j16
|  0  571.68s  9762.43s  523.74s  3091572KB   bash  -c  make  test-proof        -j10
```

The summary is that the Java and LLVM backends perform within a few percent on all numbers on both. But the OCaml backend has a 16% speedup! Wow. @dwightguth do you think it's more likely that either (1) the OCaml backend performs poorly on the original semantics in a way the LLVM backend already optimizes, or (2) the LLVM backend could be made to also see a 15% performance improvement here? I personally think (1) is more likely.

**Comment by MrChico** [comment]
Addressed formatting issues. N.B. that the speedup I'm concerned with is mainly wrt state logging.

**Comment by ehildenb** [comment]
Alright, measured klab performance this time too. Note this only runs 2 proofs with KLab:

```
* 06c521b5 - evm.md: Remove #callWithCode, dont copy the entire program to K cell             - Martin Lundfall (HEAD -> tested-profile, mrchico/omitProgram, omitProgram)
| testset: profile - 1
| 0  0.11s    0.06s     0.04s    16060KB     make  clean
| 0  0.03s    0.00s     0.03s    3324KB      rm    -rf               /mnt/ramdisk/klab
| 0  128.18s  215.94s   12.04s   837696KB    make  llvm-deps         -B
| 0  54.29s   415.87s   15.71s   2104540KB   make  build-java        build-llvm         build-ocaml                  -B  -j6
| 0  0.13s    0.10s     0.03s    15996KB     make  split-tests       -B
| 0  234.11s  2122.43s  884.27s  4488672KB   make  test-conformance  -j16               TEST_CONCRETE_BACKEND=ocaml
| 0  143.39s  1694.85s  464.09s  15365192KB  make  test-conformance  -j16               TEST_CONCRETE_BACKEND=llvm
| 0  574.93s  9721.01s  540.59s  2937952KB   make  test-proof        -j10
| 0  533.71s  760.35s   15.85s   3749512KB   make  test-klab-prove   -j4
| testset: profile - 2
| 0  0.11s    0.06s     0.05s    16008KB     make  clean
| 0  0.03s    0.00s     0.02s    3504KB      rm    -rf               /mnt/ramdisk/klab
| 0  128.73s  216.61s   11.98s   818812KB    make  llvm-deps         -B
| 0  55.00s   419.24s   16.95s   2133460KB   make  build-java        build-llvm         build-ocaml                  -B  -j6
| 0  0.13s    0.11s     0.02s    15864KB     make  split-tests       -B
| 0  233.56s  2122.06s  885.49s  4488856KB   make  test-conformance  -j16               TEST_CONCRETE_BACKEND=ocaml
| 0  143.62s  1696.84s  464.42s  15365092KB  make  test-conformance  -j16               TEST_CONCRETE_BACKEND=llvm
| 0  572.36s  9893.02s  514.91s  3194132KB   make  test-proof        -j10
| 0  531.64s  752.27s   18.77s   3715664KB   make  test-klab-prove   -j4
| testset: profile - 3
| 0  0.10s    0.07s     0.03s    15908KB     make  clean
| 0  0.03s    0.00s     0.03s    3500KB      rm    -rf               /mnt/ramdisk/klab
| 0  128.42s  215.53s   12.13s   721412KB    make  llvm-deps         -B
| 0  55.04s   419.88s   16.41s   2222712KB   make  build-java        build-llvm         build-ocaml                  -B  -j6
| 0  0.13s    0.09s     0.04s    15880KB     make  split-tests       -B
| 0  233.67s  2120.12s  886.96s  4488636KB   make  test-conformance  -j16               TEST_CONCRETE_BACKEND=ocaml
| 0  143.42s  1697.35s  464.78s  15365104KB  make  test-conformance  -j16               TEST_CONCRETE_BACKEND=llvm
| 0  566.09s  9562.15s  533.63s  3444584KB   make  test-proof        -j10
| 0  525.56s  737.87s   16.59s   3713608KB   make  test-klab-prove   -j4
| 
* 39c92c4e - deps/k: update submodule (#383)                                                  - rv-jenkins
| testset: profile - 1
| 0  0.09s    0.06s     0.02s    15944KB     make  clean
| 0  0.02s    0.00s     0.02s    3388KB      rm    -rf               /mnt/ramdisk/klab
| 0  128.05s  214.16s   12.29s   838528KB    make  llvm-deps         -B
| 0  54.33s   408.45s   15.97s   2191880KB   make  build-java        build-llvm         build-ocaml                  -B  -j6
| 0  0.13s    0.10s     0.03s    15952KB     make  split-tests       -B
| 0  276.84s  2270.87s  891.10s  4488536KB   make  test-conformance  -j16               TEST_CONCRETE_BACKEND=ocaml
| 0  142.40s  1679.35s  464.03s  15365336KB  make  test-conformance  -j16               TEST_CONCRETE_BACKEND=llvm
| 0  574.02s  9779.07s  528.77s  3276448KB   make  test-proof        -j10
| 0  523.43s  744.99s   18.44s   3732508KB   make  test-klab-prove   -j4
| testset: profile - 2
| 0  0.10s    0.06s     0.03s    16048KB     make  clean
| 0  0.02s    0.00s     0.02s    3520KB      rm    -rf               /mnt/ramdisk/klab
| 0  128.82s  216.44s   12.34s   819284KB    make  llvm-deps         -B
| 0  55.95s   417.98s   17.62s   2167300KB   make  build-java        build-llvm         build-ocaml                  -B  -j6
| 0  0.14s    0.10s     0.03s    16000KB     make  split-tests       -B
| 0  278.26s  2281.58s  895.00s  4488644KB   make  test-conformance  -j16               TEST_CONCRETE_BACKEND=ocaml
| 0  142.52s  1683.59s  462.41s  15364732KB  make  test-conformance  -j16               TEST_CONCRETE_BACKEND=llvm
| 0  554.32s  9445.97s  513.70s  3053388KB   make  test-proof        -j10
| 0  528.86s  755.55s   18.87s   3744800KB   make  test-klab-prove   -j4
| testset: profile - 3
| 0  0.11s    0.06s     0.05s    15860KB     make  clean
| 0  0.03s    0.00s     0.02s    3500KB      rm    -rf               /mnt/ramdisk/klab
| 0  128.08s  215.50s   12.27s   816492KB    make  llvm-deps         -B
| 0  55.53s   424.06s   16.56s   2175988KB   make  build-java        build-llvm         build-ocaml                  -B  -j6
| 0  0.12s    0.08s     0.04s    15952KB     make  split-tests       -B
| 0  279.06s  2287.44s  893.95s  4488608KB   make  test-conformance  -j16               TEST_CONCRETE_BACKEND=ocaml
| 0  143.13s  1690.58s  464.68s  15365020KB  make  test-conformance  -j16               TEST_CONCRETE_BACKEND=llvm
| 0  579.51s  9824.37s  536.06s  3060032KB   make  test-proof        -j10
| 0  527.46s  750.06s   18.68s   3720180KB   make  test-klab-prove   -j4
```

Still see the 16% speedup with the OCaml backend, but no appreciable difference with the klab proofs. Note that it's removing the entire `KLAB_OUT` directory every run, so that we aren't benefiting from cache on subsequent runs.

If you have a proof that it does measurably better on, I'll add the K spec to our test-set.

The OCaml backend improvement is a compelling argument for merging. Still not sure how I feel about the changes though, I'll need to look more closely on Monday.

**Comment by MrChico** [comment]
It's less LOC now

**Comment by ehildenb** [comment]
Closing this because the migration to `ByteArray` for storing the program has resulted in pretty drastic merge conflicts. If we still think it will be a performance boon, we can re-open.

**Pull request closed without merging**